### ISSUE TITLE
proposal: golang plugin hub

### ISSUE THREAD

**Issue by doujiang24** [envoyproxy/envoy#26003]
motivation: 
benefits from the golang ecosystem, people could easily implement plugins, for more long tail scenes.
we could build a plugin hub, so that people could contribute plugins to the hub, and reuse plugins.
similar to the traefik: https://plugins.traefik.io/plugins

technical details:
1. each plugin is a golang module, with a valid go.mod.
2. each plugin could declare its own config proto, default to TypedStruct.
3. people could push config through control-plane, i.e. EnvoyFilter for istio.
4. furthermore, the hub could provide a way to combine plugins into a big plugin, to reduce cgo calls which might be a bit heavy cost.

**Comment by wbpcode** [comment]
Will this a new independent repo?

**Comment by phlax** [comment]
this seems like a good idea to me - i think we should host it under the envoyproxy project

@doujiang24 i think we can get started with this in a separate repo and existing example plugins and then transfer it, add some docs, etc, once it has grown a bit

cc @mattklein123

**Comment by doujiang24** [comment]
yep, a separate repo should be good enough for starting, and we could transfer it when it is ready to ship.

**Comment by mattklein123** [comment]
I like the idea of a hub but I think it should stay a totally separate project. SGTM to host with this org.

**Comment by phlax** [comment]
somewhat relatedly - im wondering about the "canonical" way to compile a plugin - with the example we use envoy build container which is quite a heavy toolchain to add - iiuc we can compile with just the stream api and deps - maybe i dont uc tho

i mention because i guess that would be a kinda first step for a repo - some CI that tests plugins using the "canonical" build method, and also in terms of docs/contribs we probably want to publish the easiest way possible to get started

**Comment by doujiang24** [comment]
yep, it only depends on the Go compiler and the C compiler (since cgo is enabled) to compile a Go plugin, so we don't need the envoy build image to build a Go plugin. i.e. we could use the golang docker image.

**Comment by zengyuxing007** [comment]
Looking forward to it, so envoyproxy's ecology can be more rich up！

**Comment by doujiang24** [comment]
This is under early design. Please add no-stale label, thanks! cc @phlax

**Comment by jizhuozhi** [comment]
A simple idea, we can use OCI image to distribute envoy go filter just like distributing wasm artifacts.

In this way, we can easily reuse the existing cloud native infrastructure, especially the service registry running in the intranet environment. This way the hub can be just a content platform, with high quality plugins selected by the community, just like Pull Requests
